### PR TITLE
Adds Pixel Shifting

### PR DESCRIPTION
**Pull request by Zonespace27**
[Feature] [Input Wanted]
## About the PR 
This PR adds pixel shifting to the game! Pixel shifting is a thing that allows you to change the position of your character on the tile you're on, which is a purely visual change that doesn't affect your _actual_ position.

The default setup is holding U and moving via WASD to pixel shift.

Edit: I've changed it to a maximum of 14 pixels in any one direction, and a max of 22 pixels shifted from your original position (no total limit prior)

## Why's this needed? 
More ways to express yourself in RP!

[Video Example](https://imgur.com/a/1Z5K13w)

image

image

## Changelog

```changelog
(u)Zonespace
(*)Added Pixel Shifting, to use, hold U and hold W/A/S/D.
```

### CODE REVIEW

File: _std/defines/input.dm [hunk 1]
```diff
@@ -1,19 +1,21 @@
...
-#define KEY_RIGHT     0x0008
-#define KEY_RUN       0x0010
-#define KEY_THROW     0x0020
-#define KEY_POINT     0x0040
-#define KEY_EXAMINE   0x0080
-#define KEY_PULL      0x0100
-#define KEY_OPEN      0x0200
-#define KEY_BOLT      0x0400
-#define KEY_SHOCK     0x0800
+// this used to use hexadecimal, now we can have 32, fun
```
Comment (ZeWaka) [inline _std/defines/input.dm:2]: that's not how bitflags work.
we're limited to 24.

Comment (Zonespace27) [reply _std/defines/input.dm:2]: My mistake, thought it was 32

**Comment by pali6** [comment]
This will look jank mostly as we aren't on SIDE_MAP map mode. If you currently overlap with objects on other tiles things won't layer properly. We can't switch to said map mode until Lummox implements a feature I requested some time ago because currently that map mode forcibly disables the blackness plane which is necessary for how we render some things.

**Comment by Kyle2143** [comment]
Kinda wanna see video of what this would look like in practice. I assume it shouldn't be able to be done while walking?

**Comment by Zonespace27** [comment]
>I assume it shouldn't be able to be done while walking? Does it snap you back to to 0,0 pixel position when you move tiles?

Yes, you can't do it while moving, and yes it snaps you back to 0, 0

**Comment by Zonespace27** [comment]
Played with the numbers a bit, made it harder to shove yourself nearly off your tile visually

**Comment by Yellow-Mushroom** [comment]
Could you give an example for a use of this?

**Comment by Zonespace27** [comment]
> Could you give an example for a use of this?
Sure, here's a few:
- Getting up in people's faces
- Getting away from people's faces
- Leaning on tables
- Leaning on walls
- Minute motions (shifting side to side, for example)

**Comment by ZeWaka** [comment]
Can we get a visual example of shifting over a player's sprite? Or over other objects?

**Comment by Zonespace27** [comment]
> Can we get a visual example of shifting over a player's sprite? Or over other objects?

image
image
image

**Comment by Sovexe** [comment]
have concerns for interaction with projectiles, transformations, interaction with other offsets

moving just sets the offset back to 0,0 which is not always correct. Doesn't take into account existing mob offsets or bounds for larger sprites.

**Comment by Zonespace27** [comment]
> moving just sets the offset back to 0,0 which is not always correct. Doesn't take into account existing mob offsets or bounds for larger sprites.

@Sovexe Corrected by making it take a person's pre-shift x/y when they start shifting, and will reset them to that when forced out of shifting

**Comment by Sovexe** [comment]
This will help with the bugs I listed, though on the opposite end of the spectrum but we may have the problem of people saving an offset to abuse later

For example, pixel shifting mid jumpy gene proc to lock their sprite high in the air just to give an example off the top of my head

**Comment by Zonespace27** [comment]
> to give an example off the top of my head, pixel shifting mid jumpy gene proc to lock their sprite high in the air

This is pretty much impossible, because pixel shifting at all will register you to a signal on Move() to reset your position to your original.

**Comment by Zonespace27** [comment]
Closing pending me getting more time in my life and locating a better way of doing this; will reopen eventually, probably

**Pull request closed without merging**